### DISCUSSION TITLE
asyncio version of ntptime?

### DISCUSSION THREAD

**Discussion by bulletmark** [micropython/micropython#12967] [Libraries & Drivers]
Does anybody know of an implementation of `ntptime` for asyncio? i.e. so I can:

```python
await ntptime.settime()
```

**Comment by peterhinch** [comment]
Such a version would need to address a known problem. The line
```python
addr = socket.getaddrinfo(host, 123)[0][-1]
```
blocks because nobody has yet written a nonblocking DNS lookup (AFAIK of course).

**Reply by bulletmark** [reply]
I read this and thought, what does that mean my use of `uaiohttpclient()` from `micropython-lib` is actually blocking as well? So I followed it down and found this TODO in `micropython/extmod/asyncio/stream.py`:

```python
# Create a TCP stream connection to a remote host
#
# async
def open_connection(host, port):
    from errno import EINPROGRESS
    import socket

    ai = socket.getaddrinfo(host, port, 0, socket.SOCK_STREAM)[0]  # TODO this is blocking!
..
```
Doh! Although I guess this problem is lessened somewhat by the fact that host name lookups are cached at various levels.

@peterhinch do you know if there is an MP issue tracking this TODO?

**Reply by peterhinch** [reply]
> host name lookups are cached at various levels

Caching is fine when the host is fixed. I'm no expert on NTP, but I think you're supposed to do a DNS lookup on each access. A potentially different IP is retrieved from the pool each time.

I'm not aware of an issue, and a quick lookup drew a blank.

**Comment by karfas** [comment]
See e.g. [asy_udp_client](https://github.com/karfas/upy-simple-app/blob/master/lib/asy_udp_client.py) and [asy_ntp_client](https://github.com/karfas/upy-simple-app/blob/master/lib/asy_ntp_time.py)

This will block during name lookup but provides async NTP time setting. 
I'm pretty sure there are better ways to do it (especially the udp client thing), but it could be a starting point.

Edit: fixed second link.

**Reply by bulletmark** [reply]
Thanks for pushing that entire little repo up to github. I'll try those 2 files out in my app.

**Reply by peterhinch** [reply]
The second link is a 404 here.
